src/qt/rpcconsole.cpp

```diff
@@ -686,8 +687,12 @@ void RPCConsole::setClientModel(ClientModel *model)
     }
     if (!model) {
         // Client model is being set to 0, this means shutdown() is about to be called.
+        QEventLoop loop;
+        connect(&thread, &QThread::finished, &loop, &QEventLoop::quit);
         thread.quit();
-        thread.wait();
+        if (!thread.isFinished()) {
```

Comment:
Between `thread.isFinished()` and `loop.exec()` the thread can terminate and then `loop` won't quit.

Reply:
As we're implicitly using QueuedConnection, the thread finished signal won't be produced until after ```loop.exec()``` is called, that's my understanding based on the docs. Please let me know if you think I'm wrong here - I've tested this a lot, including with varying delays between ```thread.isFinished()``` and ```loop.exec()```
Another comment I wrote on this PR in a thread with @hebasto : https://github.com/bitcoin/bitcoin/pull/13674#discussion_r231021109

Let me know whether you still think this is a problem.

Reply:
You are correct, I've verified `QThread::finished` signal is emitted from a different thread which defers the `QEventLoop::quit` execution.

However, considering #15313, I'm not sure if nested event loops should be avoided. Sorry for suggesting this approach but #15313 raised some concerns in this regard.

An alternative is to expose a signal `RPCConsole::finished` to `BitcoinGUI`.

Reply:
I see the point. That's done, I've exposed the thread finished signal to ````BitcoinGUI````
Can you please review?
Thanks,